pal/src/tlsio_sl.c

```diff
@@ -0,0 +1,440 @@
+}
+
+int tlsio_sl_close(CONCRETE_IO_HANDLE tls_io,
+                        ON_IO_CLOSE_COMPLETE on_io_close_complete,
+                        void* callback_context)
+{
+    int result = 0;
+
+    if (tls_io == NULL) {
+        result = __FAILURE__;
+    }
+    else {
+        TLS_IO_INSTANCE* instance = (TLS_IO_INSTANCE*)tls_io;
+
+        if ((instance->tlsio_state == TLSIO_STATE_NOT_OPEN) ||
+            (instance->tlsio_state == TLSIO_STATE_CLOSING)) {
+            result = __FAILURE__;
+        }
+        else {
+            instance->tlsio_state = TLSIO_STATE_CLOSING;
+            instance->on_io_close_complete = on_io_close_complete;
+            instance->on_io_close_complete_context = callback_context;
+
+            close(instance->sock);
+
+            instance->tlsio_state = TLSIO_STATE_NOT_OPEN;
+            instance->on_io_close_complete(
+                                       instance->on_io_close_complete_context);
+        }
+    }
+
+    return result;
+}
+
+int tlsio_sl_send(CONCRETE_IO_HANDLE tls_io, const void* buffer, size_t size,
+                     ON_SEND_COMPLETE on_send_complete, void* callback_context)
+{
+    int result;
+
+    if (tls_io == NULL) {
+        result = __FAILURE__;
+    }
+    else {
+        TLS_IO_INSTANCE* instance = (TLS_IO_INSTANCE*)tls_io;
+
+        if (instance->tlsio_state != TLSIO_STATE_OPEN) {
+            result = __FAILURE__;
+        }
+        else {
+            const char* buf = (const char*)buffer;
+            instance->on_send_complete = on_send_complete;
+            instance->on_send_complete_callback_context = callback_context;
+
+            result = 0;
+            while (size) {
+                int res = send(instance->sock, buf, size, 0);
+                if ((res < 0) && (errno != EAGAIN)) {
+                    result = __FAILURE__;
+                    break;
+                }
+                else if (((res < 0) && (errno == EAGAIN)) || (res < size)) {
+                    /* no more space left, lets wait for more to become
+                     * available and try again.
+                     */
+                    ThreadAPI_Sleep(10);
+                }
+                if (res > 0) {
+                    size -= res;
+                    buf += res;
+                }
+            }
+            IO_SEND_RESULT oresult = result == __FAILURE__ ? IO_SEND_ERROR :
+                                                             IO_SEND_OK;
+            instance->on_send_complete(
+                          instance->on_send_complete_callback_context, oresult);
+        }
+    }
+
+    return result;
+}
+
+void tlsio_sl_dowork(CONCRETE_IO_HANDLE tls_io)
+{
+    if (tls_io != NULL) {
+        TLS_IO_INSTANCE* tls_io_instance = (TLS_IO_INSTANCE*)tls_io;
+
+        if ((tls_io_instance->tlsio_state != TLSIO_STATE_NOT_OPEN) &&
+            (tls_io_instance->tlsio_state != TLSIO_STATE_ERROR)) {
+            unsigned char buffer[64];
+            int rcv_bytes = 1;
```

Comment:
This buffer seems small and could result in a lot of repeated calls. But maybe this size is typical for our TLS implementations